### DISCUSSION TITLE
Where is the Javascript source code?

### DISCUSSION THREAD

**Discussion by paulius-valiunas** [firefly-iii/firefly-iii#8666] [Firefly III] [answered]
The repository structure is very unclear, I see a lot of minified JS files scattered all around it but I can't find any of the frontend JS.

Also, what's up with keeping your build output in git?

**Comment by JC5** [comment]
- https://github.com/firefly-iii/firefly-iii/tree/main/resources/assets/js
- https://github.com/firefly-iii/firefly-iii/tree/main/resources/assets/v2

> Also, what's up with keeping your build output in git?

Could you be a little more specific?

**Reply by paulius-valiunas** [reply]
https://github.com/firefly-iii/firefly-iii/blob/main/public/v1/js/edit_transaction.js

**Reply by paulius-valiunas** [reply]
that's just one example, I was just looking for javascript files and found _a lot_ of minified files, their names didn't look like 3rd party libraries so I assumed you used a bundler and pushed the output to github? Sorry if I made the incorrect assumption. Again, the structure could be clearer to help distinguish what is your code and what is 3rd party libraries - also, pulling those libraries from a CDN might be a good idea.

**Reply by JC5** [reply]
Yes, that's what I do. The output is saved to GitHub. This makes it a lot easier to distribute the application, which is as ready made as possible. A notable exception are the php dependencies, which users need to download themselves (they are included in the Docker container though).

Using CDNs will not be happening. I built Firefly III specifically NOT to be dependent on the internet for anything. Once installed, it runs locally. So no CDN's, scripts or telemetry from the outside world while you're using it to manage your finances.

**Reply by paulius-valiunas** [reply]
usually people use github releases and actions for this sort of thing. This avoids "oops I forgot to rebuild this file so let me make another commit that just changes a binary blob in the repo", makes the whole process more transparent (i.e. nothing is built on your local dev machine and everything can be trusted) and makes the repo lighter overall. The action can both build the source code and pull in any dependencies that are needed for the release.

**Reply by JC5** [reply]
Ok.

**Reply by JC5** [reply]
Just to be on the safe side. You know this already happens like you describe?

**Reply by paulius-valiunas** [reply]
Yes, I'm not making this up. I'm in a devops role in my job and I use a lot of open source libraries/products. The vast majority don't have any build output in git because that's considered a bad practice. They either have Github Actions set up (it's free for public repositories), or if they don't know how, someone builds the code on their dev machine and manually uploads it as a Github release and to package repositories. Either way, the build output ends up in the releases tab and doesn't clutter the git repository. And Github Actions is the preferred solution because it's fully automated and auditable.

**Reply by JC5** [reply]
I'm aware of best and worst practices, but it's always interesting to learn more about them. What I meant to say was "are you aware that *Firefly III specifically* already does this?" So I was wondering, are you curious to learn how and why Firefly III is set up the way it is? Because your feedback is welcome but I'm not seeing a question. So before I stop engaging, I just want to make sure I'm not leaving you hanging.

**Reply by paulius-valiunas** [reply]
ok, sorry for misunderstanding you, then let me repeat my question: why is build output in the git repo? Why not just leave it in the release artifact?

**Reply by paulius-valiunas** [reply]
I can see that there are some Github workflows in the repo, that's true. But I don't think it works the way I described: I see those workflows commit the build output back into the repository, which is quite the opposite of what I was trying to describe (maybe I didn't explain it well enough). What I meant is that the workflow should build the source code and publish it as an artifact, but leave the git repo alone. At least that's how it's usually done and I think it's the best practice because it leaves a clean git history with no unnecessary files in the repo.

**Reply by JC5** [reply, marked as answer]
Firefly III didn't always have generated JS as part of the app. Its inclusion is fairly new, like 5 years now. At the time, most users ran Firefly III from the git repository directly, and only installed the php dependencies using composer. This was often difficult enough for managed web hosting. It's a good idea to change this and offer artifacts only in the future.
